### PR TITLE
Week 6 login status

### PR DESCRIPTION
**Pull request by luiszaval**
- Added CSS for login button
- Added conditional CSS => if user is logged in
- Updated all code dependent on state.user.usernames => state.user.users
- Added a updateUsers action which is triggered on a new user joining or leaving socket events

### CODE REVIEW

File: src/components/SideBar/SideBar.js [hunk 1]
```diff
@@ -54,17 +54,19 @@ class SideBar extends Component {
...
                     <div key={username} className={selectedPartner && selectedPartner == username ? sidebarItemHighlightClass : "sidebar-item"}>
                         <button
                             type="button"
-                            className="sidebar-user unstyled-button"
+                            className= "sidebar-user unstyled-button"
                             value={username}
                             onClick={this.selectUser}>
                             {username}
                         </button>
+                        <div className={`login-circle ${user.logged_in ? "logged-in" : null}`}></div>
```
Comment (BitPhoenix) [inline src/components/SideBar/SideBar.js:69]: Nice!!!

File: src/modules/chat/actions.js [hunk 2]
```diff
@@ -5,11 +5,10 @@ import { actions } from "../../context";
...
     const initChat = () => (dispatch, getState) => {
         const channels = getState().channel.channels;
-        const channelIds = []
-        for (let channelId in channels) {
-            channelIds.push(channelId);
-        }
-        const usernames = getState().user.usernames;
+        const channelIds = Object.keys(channels);
+
+        const usernames = Object.keys(getState().user.users);
```
Comment (BitPhoenix) [inline src/modules/chat/actions.js:10]: Nice! Great job removing the for loop and just using Object.keys!

File: src/modules/user/actions.js [hunk 3]
```diff
@@ -15,9 +15,26 @@ const initActions = function(userService, socketService, storageService, authSer
...
 
-	const usernamesSet = actionCreator(types.SET_USERNAMES);
-	const setUsernames = (usernames) => async (dispatch) => {		
-		dispatch(usernamesSet(usernames));
+	const usersSet = actionCreator(types.SET_USERS);
+    const setUsers = (users) => (dispatch) => {
+        const usersMap = {};
+        for (let user of users) {
+            usersMap[user.username] = user;
+        }
```
Comment (BitPhoenix) [inline src/modules/user/actions.js:23]: To continue in spirit of reducing the for loops:

`const usersMap = Object.fromEntries(users.map(user => [user.username, user]));`

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries

File: src/modules/user/actions.js [hunk 4]
```diff
@@ -15,9 +15,26 @@ const initActions = function(userService, socketService, storageService, authSer
...
+        dispatch(usersSet(usersMap));
+    }
+
+	const usersUpdate = actionCreator(types.UPDATE_USERS)
+	const updateUsers = (usernameToBeUpdated) => (dispatch, getState) => {
```
Comment (BitPhoenix) [inline src/modules/user/actions.js:28]: Just need a little bit more descriptive name here for function name, something to reflect that we're marking user offline.

Perhaps it should be made more generic though, and allow for setting online/offline

`updateLoggedInStatus(username, loggedIn) `

Where true is passed in to to mark online, and false to offline

File: src/modules/user/actions.js [hunk 5]
```diff
@@ -15,9 +15,26 @@ const initActions = function(userService, socketService, storageService, authSer
...
+		const users = getState().user.users;
+		const updatedUsers = {}
+		for(let username in users){
+			if(username === usernameToBeUpdated){
+				users[username].logged_in = !users[username].logged_in
+			}
+			updatedUsers[username] = users[username]
+		}
+		dispatch(usersUpdate(updatedUsers))
```
Comment (BitPhoenix) [inline src/modules/user/actions.js:37]:
```
const user = {...users.find(user => user.username === userNameToBeUpdated)};   // finds user and creates shallow copy 
user.logged_in = !user.logged_in
```

File: src/store.js [hunk 6]
```diff
@@ -7,7 +7,7 @@ import { createLogger } from 'redux-logger';
 const configureStore = reducers => {
 	const logger = createLogger();
     const rootReducer = combineReducers(reducers);
-    const middleware = applyMiddleware(logger, thunkMiddleware);
+    const middleware = applyMiddleware(thunkMiddleware, logger);
```
Comment (BitPhoenix) [inline src/store.js:10]: Did we have this backwards before?

**Review by BitPhoenix — CHANGES_REQUESTED** [review]
Round 1: Left comments!

File: src/store.js (hunk 6 above)
Comment (luiszaval) [reply src/store.js:10]: Yeah actually. By putting it first the logger was logging both promises (all those undefined actions) and actual actions thus why our console was so clogged up. Now it should only log actions

File: src/modules/user/actions.js [hunk 7]
```diff
@@ -17,23 +17,16 @@ const initActions = function(userService, socketService, storageService, authSer
...
+	const updateLoggedInStatus = (usernameToBeUpdated, loggedIn) => (dispatch, getState) => {
 		const users = getState().user.users;
-		const updatedUsers = {}
-		for(let username in users){
-			if(username === usernameToBeUpdated){
-				users[username].logged_in = !users[username].logged_in
-			}
-			updatedUsers[username] = users[username]
-		}
+		const updatedUsers = Object.fromEntries(Object.entries(users).map(([ username, user ]) => (
```
Comment (BitPhoenix) [inline src/modules/user/actions.js:28]: Found this a little bit hard to follow, had to pause and think a lot and I think it's because of the way the update is currently being handled with map and ternary operator. I think there's a little easier way to do the update that takes less time to digest, and it also might be better to separate it out from the fromEntries / entries calls to reduce the logic chain length and make it easier to grasp in a couple of seconds what's going on 

If we're trying to update one user, can do something like this:

```
const users = {...getState().user.users}; // copy the users object  *see below*
const user = users.values().find(user => user.name === username); // find the user to update
user.logged_in = !user.logged_in; // update property on the user
dispatch(usersUpdate(users)); 
```

The only thing that has to be tweaked about this is the first line. We need a way to create a deep copy because our users object is a complex object (is a map that contains references to other user objects). One way around this is adding lodash library to the project which has simple deepclone method

https://stackoverflow.com/a/46556129

**Review by BitPhoenix — CHANGES_REQUESTED** [review]
Left round 2 comment!
